### PR TITLE
Use ENV[*] to override rbconfig.rb file

### PR DESCRIPTION
**Pull request by pierrsc**
- [x] Look for toolchain programs from environment variables first
  - [x] Test that a gem with native extensions can be installed

Checklist
* [x] Used a fork of the feedstock to propose changes
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

File: recipe/rbconfig.sed [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env sed -f
```
Comment (pierrsc) [inline recipe/rbconfig.sed]: The list of environment variables come from https://github.com/ContinuumIO/anaconda-issues/issues/9863#issuecomment-460512567

**Comment by pierrsc** [comment]
@conda-forge-admin, please rerender

**Comment by pierrsc** [comment]
Hi @conda-forge/ruby team,

Would it be possible to review this PR that rewrites the rbconfig.rb file to support environment variables from supplied by a toolchain? The folks using ruby would still need to install the compiler, just like it is done currently for python. However,I believe it would close #14, close #18 , and close ContinuumIO/anaconda-issues/issues/9863. And, It should unblock conda/staged-recipes#5978.

cc: @asford, @msarahan, @lmrogriguezr, @phil-blain, @bgruening

File: recipe/rbconfig.sed [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/sed -f
+s/\(CONFIG\["\(CC\)"\] =\) \(.*\)/\1 ENV['\2'] || \3/
+s/\(CONFIG\["\(CPP\)"\] =\) \(.*\)/\1 ENV['\2'] || \3/
```
Comment (sodre) [inline recipe/rbconfig.sed]: @pierrsc, it looks like you missed CXX.

**Comment by pierrsc** [comment]
@scopatz or @conda-forge/core,  

Would you be able to take a look at the ruby-feedstock PR. It is a fix for the compiler variables in rbconfig.rb. 

cc: @lmrodriguezr

**Comment by isuruf** [comment]
Have you tried changing the variables of absolute paths to relative paths before running `configure`? This might be a better fix than the sed hack.

**Comment by pierrsc** [comment]
If it worked (which I can test), wouldn't it mean that ruby would ignore the environment variables set by our C, C++ compilers? Is that what is desired here?

**Comment by isuruf** [comment]
> If it worked (which I can test), wouldn't it mean that ruby would ignore the environment variables set by our C, C++ compilers? Is that what is desired here?

I think the issue is that ruby is remembering the full path to the compilers which don't exist after build process is done.

So instead of `CC=$PREFIX/bin/$CHOST-cc` if we had `CC=$CHOST-cc`, they'll get picked up when the compilers are installed. You can do `export CC=$(basename $CC)` to fix this

**Comment by pierrsc** [comment]
> > If it worked (which I can test), wouldn't it mean that ruby would ignore the environment variables set by our C, C++ compilers? Is that what is desired here?
> 
> I think the issue is that ruby is remembering the full path to the compilers which don't exist after build process is done.

I agree with you. My only question was only whether that is the end behavior that we want.

> 
> So instead of `CC=$PREFIX/bin/$CHOST-cc` if we had `CC=$CHOST-cc`, they'll get picked up when the compilers are installed. You can do `export CC=$(basename $CC)` to fix this

I am running a build locally now with the environment variables changed similar to how you described. Once it passed it locally, I'll push it to git. Here is what I added to `build.sh`

```bash
export CC=$(basename $CC)
export CPP=$(basename $CPP)
export CXX=$(basename $CXX)
export STRIP=$(basename $STRIP)
export NM=$(basename $NM)
export OBJDUMP=$(basename $OBJDUMP)
export AS=$(basename $AS)
export AR=$(basename $AR)
export RANLIB=$(basename $RANLIB)
export LD=$(basename $LD)
```

**Comment by pierrsc** [comment]
@isuruf, I made the modifications you requested. It should be ready to merge / re-review.

**Comment by scopatz** [comment]
Thanks!

**Comment by lmrodriguezr** [comment]
This is awesome! Thanks!

**Comment by choldgraf** [comment]
This is super useful! A naive question: is there any lag between when PRs get merged into a feedstock, and when they go live on conda forge? Any way to know when they're updated?

**Comment by lmrodriguezr** [comment]
@choldgraf You can check [here](https://anaconda.org/conda-forge/ruby/files?channel=main&type=).

It's already there. You can force the build referenced here to be installed using:
```
conda install -y -c conda-forge ruby=2.4.5=h28fbfb0_1003
```

**Comment by lmrodriguezr** [comment]
FYI, there remains a relic reference to the building environment in `rbconfig.rb` that can cause some warnings. `CONFIG["LDFLAGS"]` includes `-L/usr/local/miniconda/conda-bld/ruby_1566401986805/_build_env/lib/clang/4.0.1/lib`. This causes the warning:
```
ld: warning: directory not found for option '-L/usr/local/miniconda/conda-bld/ruby_1566401986805/_build_env/lib/clang/4.0.1/lib'
```

This doesn't seem to be fatal to compilations, AFAICT, so it's probably safe to ignore.

**Pull request merged**